File: src/Haskell/Hamt/src/Durable7/Hamt/PersistentDirectedGraph.hs
-- | A persistent directed graph over hashed vertices.
--
-- Successors and predecessors are both indexed, so neither direction is a scan. Every operation
-- returns a new version and leaves its inputs valid, sharing unchanged structure, so an edit
-- copies a path rather than the whole collection.
module Durable7.Hamt.PersistentDirectedGraph
  ( PersistentDirectedGraph
  , empty
  , emptyWith
  , fromLists
  , vertexCount
  , edgeCount
  , null
  , policy
  , memberVertex
  , memberEdge
  , actualVertex
  , successors
  , predecessors
  , outDegree
  , inDegree
  , insertVertex
  , insertEdge
  , deleteEdge
  , deleteVertex
  , clear
  , reverse
  , vertices
  , edges
  , validStructure
  ) where

import Prelude hiding (null, reverse)

import qualified Data.List as List
import Durable7.Hamt.Hashable (Hashable)
import Durable7.Hamt.HashMap (HashPolicy)
import qualified Durable7.Hamt.HashMap as HashMap
import Durable7.Hamt.HashSet (HashSet)
import qualified Durable7.Hamt.HashSet as HashSet
import Durable7.Hamt.Relation (Relation)
import qualified Durable7.Hamt.Relation as Relation

-- | A persistent directed graph over hashed vertices. Successors and predecessors are both indexed,
-- so neither direction is a scan.
data PersistentDirectedGraph v = PersistentDirectedGraph !(HashSet v) !(Relation v v)

-- | The empty graph.
empty :: (Eq v, Hashable v) => PersistentDirectedGraph v
empty = emptyWith HashMap.defaultPolicy

-- | The empty graph under the given policy, which it retains.
emptyWith :: HashPolicy v -> PersistentDirectedGraph v
emptyWith verticesPolicy =
  PersistentDirectedGraph (HashSet.emptyWith verticesPolicy) (Relation.emptyWith verticesPolicy verticesPolicy)

-- | A graph holding the given vertices and edges.
fromLists :: HashPolicy v -> [v] -> [(v, v)] -> PersistentDirectedGraph v
fromLists verticesPolicy initialVertices initialEdges =
  List.foldl' (\graph (source, target) -> insertEdge source target graph)
    (List.foldl' (flip insertVertex) (emptyWith verticesPolicy) initialVertices)
    initialEdges

-- | Number of vertices.
vertexCount :: PersistentDirectedGraph v -> Int
vertexCount (PersistentDirectedGraph storedVertices _) = HashSet.size storedVertices

-- | Number of directed edges.
edgeCount :: PersistentDirectedGraph v -> Int
edgeCount (PersistentDirectedGraph _ storedEdges) = Relation.size storedEdges

-- | Whether the graph holds no vertices.
null :: PersistentDirectedGraph v -> Bool
null graph = vertexCount graph == 0

-- | The policy the graph retains.
policy :: PersistentDirectedGraph v -> HashPolicy v
policy (PersistentDirectedGraph storedVertices _) = HashSet.policy storedVertices

-- | Whether the vertex is present.
memberVertex :: v -> PersistentDirectedGraph v -> Bool
memberVertex vertex (PersistentDirectedGraph storedVertices _) = HashSet.member vertex storedVertices

-- | Whether the directed edge is present.
memberEdge :: v -> v -> PersistentDirectedGraph v -> Bool
memberEdge source target (PersistentDirectedGraph _ storedEdges) = Relation.member source target storedEdges

-- | The stored vertex representative.
actualVertex :: v -> PersistentDirectedGraph v -> Maybe v
actualVertex vertex (PersistentDirectedGraph storedVertices _) = HashSet.actualValue vertex storedVertices

-- | The vertices the given vertex points at.
successors :: v -> PersistentDirectedGraph v -> HashSet v
successors source graph@(PersistentDirectedGraph _ storedEdges) =
  maybe (HashSet.emptyWith (policy graph)) id (Relation.lookupRights source storedEdges)

-- | The vertices pointing at the given vertex. Maintained as its own index, so this is a lookup
-- rather than a scan.
predecessors :: v -> PersistentDirectedGraph v -> HashSet v
predecessors target graph@(PersistentDirectedGraph _ storedEdges) =
  maybe (HashSet.emptyWith (policy graph)) id (Relation.lookupLefts target storedEdges)

-- | How many edges leave the vertex.
outDegree :: v -> PersistentDirectedGraph v -> Int
outDegree source = HashSet.size . successors source

-- | How many edges enter the vertex.
inDegree :: v -> PersistentDirectedGraph v -> Int
inDegree target = HashSet.size . predecessors target

-- | A graph containing the vertex, with no edges added.
insertVertex :: v -> PersistentDirectedGraph v -> PersistentDirectedGraph v
insertVertex vertex graph@(PersistentDirectedGraph storedVertices storedEdges)
  | HashSet.member vertex storedVertices = graph
  | otherwise = PersistentDirectedGraph (HashSet.insert vertex storedVertices) storedEdges

-- | Inserts an edge and implicitly inserts missing endpoint vertices.
insertEdge :: v -> v -> PersistentDirectedGraph v -> PersistentDirectedGraph v
insertEdge source target graph@(PersistentDirectedGraph storedVertices storedEdges)
  | Relation.member source target storedEdges = graph
  | otherwise =
      let withSource = HashSet.insert source storedVertices
          withEndpoints = HashSet.insert target withSource
          actualSource = requiredRepresentative source withEndpoints
          actualTarget = requiredRepresentative target withEndpoints
       in PersistentDirectedGraph withEndpoints (Relation.insert actualSource actualTarget storedEdges)

-- | A graph without that directed edge, leaving both endpoints in place.
deleteEdge :: v -> v -> PersistentDirectedGraph v -> PersistentDirectedGraph v
deleteEdge source target graph@(PersistentDirectedGraph storedVertices storedEdges)
  | not (Relation.member source target storedEdges) = graph
  | otherwise = PersistentDirectedGraph storedVertices (Relation.delete source target storedEdges)

-- | A graph without the vertex and without any edge touching it.
deleteVertex :: v -> PersistentDirectedGraph v -> PersistentDirectedGraph v
deleteVertex vertex graph@(PersistentDirectedGraph storedVertices storedEdges)
  | not (HashSet.member vertex storedVertices) = graph
  | otherwise =
      PersistentDirectedGraph
        (HashSet.delete vertex storedVertices)
        (Relation.deleteRight vertex (Relation.deleteLeft vertex storedEdges))

-- | The empty graph, retaining the same policies.
clear :: PersistentDirectedGraph v -> PersistentDirectedGraph v
clear graph
  | null graph = graph
  | otherwise = emptyWith (policy graph)

-- | The graph in the opposite order.
reverse :: PersistentDirectedGraph v -> PersistentDirectedGraph v
reverse (PersistentDirectedGraph storedVertices storedEdges) =
  PersistentDirectedGraph storedVertices (Relation.inverse storedEdges)

-- | The vertices.
vertices :: PersistentDirectedGraph v -> [v]
vertices (PersistentDirectedGraph storedVertices _) = HashSet.toList storedVertices

-- | The directed edges.
edges :: PersistentDirectedGraph v -> [(v, v)]
edges (PersistentDirectedGraph _ storedEdges) = Relation.toList storedEdges

-- | Whether the graph's structural invariants hold. For tests and diagnostics.
validStructure :: PersistentDirectedGraph v -> Bool
validStructure graph@(PersistentDirectedGraph _ storedEdges) =
  Relation.validStructure storedEdges
    && all (\(source, target) -> memberVertex source graph && memberVertex target graph) (Relation.toList storedEdges)

requiredRepresentative :: v -> HashSet v -> v
requiredRepresentative value values = case HashSet.actualValue value values of
  Just actual -> actual
  Nothing -> error "PersistentDirectedGraph failed to retain an inserted vertex"

File: src/Kotlin/Hamt/src/durable7/hamt/PersistentDirectedGraph.kt
/*
 * Persistent directed graph with explicit vertices and both adjacency directions.
 *
 * Keeping the vertex set explicit makes an isolated vertex representable, and removing a vertex
 * removes its incident edges rather than leaving the edge relation naming a vertex that no longer
 * exists.
 */
package durable7.hamt

/** One directed edge, from a source vertex to a target vertex. */
public data class DirectedEdge<V>(public val source: V, public val target: V)

/** Vertex and edge counts returned by a successful structural audit. */
public data class PersistentDirectedGraphStatistics(
    public val vertexCount: Int,
    public val edgeCount: Long,
)

/** Immutable directed graph with explicit vertices and bidirectionally indexed edges. */
public class PersistentDirectedGraph<V> private constructor(
    private val storedVertices: PersistentHashSet<V>,
    private val storedEdges: PersistentRelation<V, V>,
) : Iterable<V> {
    public companion object {
        /** An empty graph. One policy governs vertex identity in both directions. */
        public fun <V> empty(policy: HashPolicy<V> = defaultHashPolicy()): PersistentDirectedGraph<V> =
            PersistentDirectedGraph(PersistentHashSet.empty(policy), PersistentRelation.empty(policy, policy))

        /**
         * A graph with every vertex and edge. An edge naming a vertex not in [vertices] adds that vertex, so the
         * result is always well formed.
         */
        public fun <V> from(
            vertices: Iterable<V>,
            edges: Iterable<Pair<V, V>>,
            policy: HashPolicy<V> = defaultHashPolicy(),
        ): PersistentDirectedGraph<V> {
            var result = empty<V>(policy)
            for (vertex in vertices) result = result.addVertex(vertex)
            for ((source, target) in edges) result = result.addEdge(source, target)
            return result
        }
    }

    /** The hash policy governing vertex identity. */
    public val policy: HashPolicy<V> get() = storedVertices.policy
    /**
     * Number of vertices, including isolated ones. Vertices exist independently of edges here, which is what lets
     * the graph represent an isolated node at all.
     */
    public val vertexCount: Int get() = storedVertices.size
    /** Number of directed edges. */
    public val edgeCount: Long get() = storedEdges.pairCount
    /** Whether the graph has no vertices. A graph with vertices but no edges is not empty. */
    public val isEmpty: Boolean get() = storedVertices.isEmpty
    /** The vertices, in the vertex set's iteration order. */
    public val vertices: Sequence<V> get() = storedVertices.asSequence()
    /** The directed edges, each as a source-target pair. */
    public val edges: Sequence<DirectedEdge<V>> get() = storedEdges.asSequence().map { DirectedEdge(it.first, it.second) }
    /**
     * The graph with every edge reversed. O(1): the two directions of the underlying relation are already indexed,
     * so this swaps them rather than rebuilding.
     */
    public val reversed: PersistentDirectedGraph<V> get() = PersistentDirectedGraph(storedVertices, storedEdges.inverse())

    /** Whether [vertex] is present. */
    public fun containsVertex(vertex: V): Boolean = storedVertices.contains(vertex)
    /**
     * Whether a directed edge runs from [source] to [target]. Direction matters: this says nothing about the
     * reverse edge.
     */
    public fun containsEdge(source: V, target: V): Boolean = storedEdges.contains(source, target)
    /** The *stored* vertex representative equal to [vertex], or `null` when absent. */
    public fun actualVertex(vertex: V): V? = storedVertices.get(vertex)
    /**
     * The vertices reachable by one outgoing edge from [vertex]. An empty set both when the vertex has no
     * successors and when it is not in the graph.
     */
    public fun successors(vertex: V): PersistentHashSet<V> =
        storedEdges.rightsFor(vertex) ?: PersistentHashSet.empty(policy)
    /**
     * The vertices with one incoming edge into [vertex]. Costs the same as [successors], since both directions are
     * indexed.
     */
    public fun predecessors(vertex: V): PersistentHashSet<V> =
        storedEdges.leftsFor(vertex) ?: PersistentHashSet.empty(policy)
    /** How many edges leave [vertex]. */
    public fun outDegree(vertex: V): Int = successors(vertex).size
    /** How many edges enter [vertex]. */
    public fun inDegree(vertex: V): Int = predecessors(vertex).size

    /**
     * A graph containing [vertex]. Returns the receiver unchanged when it is already present, keeping the stored
     * representative.
     */
    public fun addVertex(vertex: V): PersistentDirectedGraph<V> {
        val updated = storedVertices.put(vertex)
        return if (updated === storedVertices) this else PersistentDirectedGraph(updated, storedEdges)
    }

    /** Adds an edge and implicitly adds missing endpoint vertices. */
    public fun addEdge(source: V, target: V): PersistentDirectedGraph<V> {
        if (storedEdges.contains(source, target)) return this
        val nextVertices = storedVertices.put(source).put(target)
        val actualSource = checkNotNull(nextVertices.get(source))
        val actualTarget = checkNotNull(nextVertices.get(target))
        return PersistentDirectedGraph(nextVertices, storedEdges.add(actualSource, actualTarget))
    }

    /**
     * A graph without the edge from [source] to [target]. Both endpoints remain as vertices. Returns the receiver
     * unchanged when the edge is absent.
     */
    public fun removeEdge(source: V, target: V): PersistentDirectedGraph<V> {
        val nextEdges = storedEdges.remove(source, target)
        return if (nextEdges === storedEdges) this else PersistentDirectedGraph(storedVertices, nextEdges)
    }

    /**
     * A graph without [vertex] and every edge touching it, in either direction. Returns the receiver unchanged when
     * the vertex is absent.
     */
    public fun removeVertex(vertex: V): PersistentDirectedGraph<V> {
        if (!storedVertices.contains(vertex)) return this
        return PersistentDirectedGraph(
            storedVertices.remove(vertex),
            storedEdges.removeLeft(vertex).removeRight(vertex),
        )
    }

    /** An empty graph retaining the same policy. */
    public fun clear(): PersistentDirectedGraph<V> = if (isEmpty) this else empty(policy)

    /**
     * Walk the whole graph and check its invariants - both directions of the edge relation agreeing, and every edge
     * endpoint present as a vertex - returning its shape measurements. A defensive audit for tests, not a routine
     * operation.
     *
     * @throws IllegalStateException on the first violation found.
     */
    public fun validateStructure(): PersistentDirectedGraphStatistics {
        storedEdges.validateStructure()
        for (edge in storedEdges) {
            check(storedVertices.contains(edge.first) && storedVertices.contains(edge.second)) {
                "PersistentDirectedGraph edge endpoint is absent from the vertex set."
            }
        }
        return PersistentDirectedGraphStatistics(vertexCount, edgeCount)
    }

    override fun iterator(): Iterator<V> = storedVertices.iterator()
}
